File: src/set49mode.cpp
/*
    set49mode sets the mode of Groovy Game Gear's 49-way Joystick. 
    
    This script is based off of De Waegeneer Gijsbrecht's 2018 SetUltrastik360 script

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
    See the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.

    contact: mahuti@gmail.com
 */

#include <chrono>
#include <cstring>
#include <iostream>
#include <libusb-1.0/libusb.h>
#include <sstream>
#include <thread>
#include <vector>
#include <algorithm>

#define UM_REQUEST_TYPE 0x21
#define UM_REQUEST 9
#define UM_TIMEOUT 2000
#define J_VENDOR_ID 0xFAFA
#define J_REPORT 0x0200
#define J_MESG_LENGTH 2
#define J_INTERFACE 0
#define KERNEL_DRIVER_ATTACHED 1
#define VERSION "1.0.0"

void cleanup(libusb_context * context, libusb_device_handle * devicehandle  ) {
   if (devicehandle) {
      libusb_close(devicehandle);
   }
   if (context) {
      libusb_exit(context);
   }
}

[[noreturn]] void
   errorhandler(libusb_context * context, libusb_device_handle * devicehandle, const std::string & errorMessage) {
   std::cerr << "ERROR: " << errorMessage << "\n";
   cleanup(context, devicehandle);
   exit(EXIT_FAILURE);
}

[[noreturn]] void errorhandler(libusb_context * context, libusb_device_handle * devicehandle, int rc) {
   std::cerr << "ERROR: " << libusb_error_name(rc) << " - " << libusb_strerror(static_cast<libusb_error>(rc)) << "\n";
   cleanup(context, devicehandle);
   exit(EXIT_FAILURE);
}

auto apply49mode(unsigned char modeId, unsigned char product_id) {
   libusb_context *       context(nullptr);
   libusb_device *        device(nullptr);
   libusb_device_handle * devicehandle(nullptr);
   libusb_device **       devices;
   auto                   rc = libusb_init(&context);

   if (rc != LIBUSB_SUCCESS) {
      errorhandler(context, devicehandle, rc);
   }
#if LIBUSB_API_VERSION >= 0x01000106
   libusb_set_option(context, LIBUSB_OPTION_LOG_LEVEL, LIBUSB_LOG_LEVEL_WARNING);
#else
   libusb_set_debug(context, LIBUSB_LOG_LEVEL_WARNING);
#endif
   auto deviceCount = libusb_get_device_list(context, &devices);
   
     for (auto deviceIndex(0); deviceIndex < deviceCount; deviceIndex++) {
      device = devices[deviceIndex];
      libusb_device_descriptor descriptor = {};
      rc = libusb_get_device_descriptor(device, &descriptor);
      if (rc != LIBUSB_SUCCESS) {
         std::cout << "WARNING: " << libusb_error_name(rc) << " - " << libusb_strerror(static_cast<libusb_error>(rc))
                   << " - trying to proceed...\n";
      } else {
         if ((descriptor.idVendor == J_VENDOR_ID) && (descriptor.idProduct == product_id)) {
            break;
         }
      }
      device = nullptr;
   }
  
   if (device == nullptr) {
      
       std::cout << "GPWiz49Way Device id# " << static_cast<unsigned>(product_id) << " not found\n";
       cleanup(context, devicehandle);
       return 0; 
      // errorhandler(context, devicehandle, ss.str());
   } else {   // good to go
      rc = libusb_open(device, &devicehandle);
      if (rc != LIBUSB_SUCCESS) {
         errorhandler(context, devicehandle, rc);
      }
      if (devicehandle) {
         if (libusb_kernel_driver_active(devicehandle, J_INTERFACE) == KERNEL_DRIVER_ATTACHED) {
         
            rc = libusb_detach_kernel_driver(devicehandle, J_INTERFACE);
            libusb_claim_interface(devicehandle,J_INTERFACE);
            
            if (rc != LIBUSB_SUCCESS) {
               errorhandler(context, devicehandle, rc);
            }
         }
         
         
         unsigned char message[J_MESG_LENGTH] = { 204, modeId };  

        rc = libusb_control_transfer(devicehandle,
                                      UM_REQUEST_TYPE,
                                      UM_REQUEST,
                                      J_REPORT,
                                      J_INTERFACE,
                                      message,
                                      J_MESG_LENGTH,
                                      UM_TIMEOUT);
         std::cout << "GPWiz49way Device id# " << static_cast<unsigned>(product_id) << ((rc == sizeof(message)) ? " successfully converted " : " failed to convert ") << "to mode# " << static_cast<unsigned>(modeId) << "\n";
      }
   }
   libusb_release_interface(devicehandle, J_INTERFACE);
   rc = libusb_attach_kernel_driver(devicehandle, J_INTERFACE);
   libusb_free_device_list(devices, 1);
   cleanup(context, devicehandle);
   return 0; 
}

int main(int argc, char *argv[]) {
  unsigned char modeId(1);
  long int joystick(0);
  switch (argc) {
      case 3:
        try {
            joystick = std::stol(argv[2], nullptr, 10);
            if (joystick < 0 || joystick > 1) {
                joystick = 0;
            }
        } catch (std::exception &) {
            joystick = 0;
        }
      case 2:
        try {
            modeId = std::stol(argv[1], nullptr, 10);
            if (modeId < 1 || modeId > 8) {
                modeId = 1;
            }
        } catch (std::exception &) {
            modeId = 1;
        }

    break;
  default:
  
  
   std::cout << "GPWiz49 switcher Version " << VERSION << "\n\n"
         << "This software is to set the joystick mode for the GPWiz49 Joystick from GroovyGameGear. \n"
        << "https://groovygamegear.com/webstore/index.php?main_page=product_info&cPath=76_81&products_id=233\n"
        << "This script will not work on the other GGG 49way encoders, just the GPWiz49,\n"
        << "though it should be easy to change the product ID to support other encoders\n"
        << "\n"
        << "=====================================\n"
        << "\n"
        << "Note: To use the GGG GPWiz49 on a Raspberry Pi, you will need to set some HID quirks.\n" 
        << "1. Create file /etc/modprobe.d/usbhid.conf with the following:\n"
        << "    options usbhid quirks=0xFAFA:0x0007:0x00000020,0xFAFA:0x0008:0x00000020\n"
        << "2. Add a hid quirk to the end of boot script: sudo pico /boot/cmdline.txt\n"
        << "    usbhid.quirks=0xFAFA:0x0007:0x00000020,0xFAFA:0x0008:0x00000020\n"
        << "If you want to use one of the other GGG products, you will need to change the existing product id: 0x0007\n"
        << "3. You will also need to add a UDEV rule so that this app can be run without ROOT privileges\n"
        << "    sudo pico :/etc/udev/rules.d/50-set49mode.rules\n" 
        << "Add this:\n" 
        << "    ACTION=='add', SUBSYSTEMS=='usb', ATTRS{idVendor}=='fafa', ATTRS{idProduct}=='00ff', MODE:='666' \n"
        << "After saving and closing the UDEV rule, Reload UDEV\n"
        << "    sudo udevadm control --reload\n"
        << "\n"
        << "====================================\n"
        << "\n"
        << "Usage\n"
        << "set49mode <joystick-mode> <joystick-type>\n"
        << "modes:\n"
        << "1: 49-way (default)\n"
        << "2: Progressive 49\n"
        << "3: 8-Way\n"
        << "4: 4-Way\n"
        << "5: 4-Way Diagonal\n"
        << "6: 2-Way Horizontal\n"
        << "7: 2-Way Vertical\n"
        << "8: 16-Way / Large dead zone in center\n"
        << "\n"
        << "joystick types:\n"
        << "0: Happs (default doesn't need to be passed in)\n"
        << "1: Williams\n"
        << "\n"
        << "examples\n"
        << "    set49mode 5 1\n"
        << "GPWiz49 will switch to 4-Way Diagonal on a Williams stick\n"
        << "     set49mode 8\n"
        << "Will switch to 16-Way on a Happs stick\n"
        << "\n"
        << "====================================\n"
        << "\n"
        << "This program comes with ABSOLUTELY NO WARRANTY. This is free software,\nand you are welcome to redistribute it under certain conditions.\n"
        << "license: GNU GENERAL PUBLIC LICENSE Version 3, 29 June 2007\nCopyright (C) 2007 Free Software Foundation, Inc. <https://fsf.org/>\n";
    return EXIT_SUCCESS;
  }
  // williams stick, just add 10. 
  if (joystick)
  {
      modeId+=10; 
  }
  
   apply49mode(modeId, 0x0007);
   apply49mode(modeId, 0x0008);
   apply49mode(modeId, 0x0009);
   apply49mode(modeId, 0x000a);

  return EXIT_SUCCESS;
}

File: python-version/set49mode.py
#!/usr/bin/python

# this is a self-contained script that does not need compiling
# 
# To use this, you must have PYUSB installed
# sudo apt-get install python libusb-1.0
# sudo apt-get install python-pip
# sudo pip install --upgrade pyusb
# call like so: 
# sudo python set49mode.py --debug --mode 1 --joystick 0
# sudo python set49mode.py -h

import usb.core
import usb.util
import sys, getopt
#import time
import os

argv = sys.argv
msg = [204,1]
mode = '1'
joystick = 'happs'
debug = 0

#find device
dev = usb.core.find(idVendor=0xfafa, idProduct=0x0007)
if dev is None:
    raise ValueError('No GPWiz49 could be found')
    exit()
    
#claim device
if dev.is_kernel_driver_active(0) is True:
    dev.detach_kernel_driver(0)
    usb.util.claim_interface(dev,0)

#activate device
if not dev.get_active_configuration():
    dev.set_configuration(1)

# happs
# 1 2041 = 49
# 2 2042 = progressive 49
# 3 2043 = 8-way
# 4 2044 = 4-way
# 5 2045 = 4-way diag
# 6 2046 = 2-way horiz
# 7 2047 = 2-way vert
# 8 2048 = 16-way
#
# williams
# 1 20411 = 49
# 2 20412 = progressive 49
# 3 20413 = 8-way
# 4 20414 = 4-way
# 5 20415 = 4-way diag
# 6 20416 = 2-way horiz
# 7 20417 = 2-way vert
# 8 20418 = 16-way

# Remove 1st argument from the list of command line arguments 
argumentList = sys.argv[1:] 
  
# Options 
options = "hdm:j:"
  
# Long options 
long_options = ["Help", "Debug", "Mode=", "Joystick="]           
    
try:
    arguments, values = getopt.getopt(argumentList, options, long_options)
except getopt.GetoptError:
    print 'set49mode.py -h -m <mode> -j <happs/williams>'
    sys.exit(2)
for currentArgument, currentValue in arguments:
    if currentArgument in ("-h", "--help"):
        print 'sudo python set49mode.py -m <mode> -j <happs/williams>'
        print ''
        print 'This software is to set the joystick mode for the GPWiz49 Joystick from GroovyGameGear. '
        print 'This script will not work on the other GGG 49way encoders, just the GPWiz 49,'
        print 'though it should be easy to change the product ID to support other encoders'
        print 'https://groovygamegear.com/webstore/index.php?main_page=product_info&cPath=76_81&products_id=233'
        print 'If no arguments are passed in, this software will set the joystick to 49-way mode for a Happs joystick'
        print ''
        print '====================================='
        print ''
        print 'Note: To use the GGG GPWiz49 on a Raspberry Pi, you will need to set some HID quirks.' 
        print '1. Create file /etc/modprobe.d/usbhid.conf with the following:'
        print 'options usbhid quirks=0xFAFA:0x0007:0x00000020,0xFAFA:0x0008:0x00000020'
        print '2. Add a hid quirk to the end of boot script: sudo pico /boot/cmdline.txt'
        print 'usbhid.quirks=0xFAFA:0x0007:0x00000020,0xFAFA:0x0008:0x00000020'
        print 'If you want to use one of the other GGG products, you will need to change the existing product id: 0x0007'
        print ''
        print '===================================='
        print ''
        print '-h --help Help'
        print '-d --debug Debug mode, will output information about selection & joystick type'
        print ''
        print '===================================='
        print ''
        print '-m --mode Restriction Mode'
        print '1: 49-way (default)'
        print '2: Progressive 49'
        print '3: 8-Way'
        print '4: 4-Way'
        print '5: 4-Way Diagonal'
        print '6: 2-Way Horizontal'
        print '7: 2-Way Vertical'
        print '8: 16-Way / Large dead zone in center'
        print ''
        print '===================================='
        print '-j --joystick Joystick'
        print 'happs (default)'
        print 'williams'
        sys.exit()
    elif currentArgument in ("-m", "--mode"):
        if debug:
            print 'mode set', currentValue
        mode = currentValue
    elif currentArgument in ("-d", "--debug"):
        debug = 1
    elif currentArgument in ("-j", "--joystick"):
        if debug:
            print 'joystick set', currentValue
        joystick = currentValue

if joystick == "happs":
    if mode == "2":
        msg = [204,2] 
    elif mode == "3": 
        msg = [204,3]
    elif mode == "4": 
        msg = [204,4]
    elif mode == "5": 
        msg = [204,5]
    elif mode == "6": 
        msg = [204, 6] 
    elif mode == "7": 
        msg = [204,7] 
    elif mode == "8": 
        msg = [204,8]
    else: 
        msg = [204,1]    
else: 
    if mode == "2":
        msg = [204,12]
    elif mode == "3": 
        msg = [204,13] 
    elif mode == "4": 
        msg = [204,14] 
    elif mode == "5": 
        msg = [204,15]
    elif mode == "6": 
        msg = [204,16] 
    elif mode == "7": 
        msg = [204,17] 
    elif mode == "8": 
        msg = [204,18] 
    else: 
        msg = [204,11]   

if debug:
    print 'Mode ', mode
    print 'Joystick ', joystick
    print 'Message ', msg
    print 'argv', argv


#HID - 0x21 - 
#SET_REPORT - 0x09
#Report Type and Report ID - 0x02 0x00
#Interface - 0
#list of values - msg

response = dev.ctrl_transfer(0x21, 0x09, 0x0200, 0, msg)
if debug:
    print "response", response

#release device
usb.util.release_interface(dev,0)
dev.attach_kernel_driver(0)


exit()
